core/trino-main/src/main/java/io/trino/split/RemoteSplit.java

```diff
@@ -76,16 +71,106 @@ public boolean isRemotelyAccessible()
     public String toString()
     {
         return toStringHelper(this)
-                .add("taskId", taskId)
-                .add("location", location)
+                .add("exchangeInput", exchangeInput)
                 .toString();
     }
 
     @Override
     public long getRetainedSizeInBytes()
     {
-        return INSTANCE_SIZE
-                + taskId.getRetainedSizeInBytes()
-                + estimatedSizeOf(location);
+        return INSTANCE_SIZE + exchangeInput.getRetainedSizeInBytes();
+    }
+
+    @JsonTypeInfo(
+            use = JsonTypeInfo.Id.NAME,
+            property = "@type")
+    @JsonSubTypes({
+            @JsonSubTypes.Type(value = DirectExchangeInput.class, name = "direct"),
+            @JsonSubTypes.Type(value = ExternalExchangeInput.class, name = "external")})
+    public interface ExchangeInput
+    {
+        long getRetainedSizeInBytes();
+    }
+
+    public static class DirectExchangeInput
+            implements ExchangeInput
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(DirectExchangeInput.class).instanceSize();
+
+        private final TaskId taskId;
+        private final String location;
+
+        @JsonCreator
+        public DirectExchangeInput(
+                @JsonProperty("taskId") TaskId taskId,
+                @JsonProperty("location") String location)
+        {
+            this.taskId = requireNonNull(taskId, "taskId is null");
+            this.location = requireNonNull(location, "location is null");
+        }
+
+        @JsonProperty
+        public TaskId getTaskId()
+        {
+            return taskId;
+        }
+
+        @JsonProperty
+        public String getLocation()
+        {
+            return location;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("taskId", taskId)
+                    .add("location", location)
+                    .toString();
+        }
+
+        @Override
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE
+                    + taskId.getRetainedSizeInBytes()
+                    + estimatedSizeOf(location);
+        }
+    }
+
+    public static class ExternalExchangeInput
```

Comment:
I remember having this discussion with somebody. The argument against calling it "Spooling" was that it implies a "hard" storage in a way. Though it can also be interpreted as a more generic concept. I don't have a strong opinion here. @losipiuk , @linzebing thoughs?

Reply:
Spooling doesn't imply hard storage. It just means a mechanism involving buffering. It could be on disk, on memory, etc.

Reply:
Both look fine to me. I kinda got used to "external" but I can live with a change.

Reply:
Let me rename it then, as we are going to talk more about the intermediate buffer as a "Spooling space"